File: modules/01-math-foundations/01-linear-algebra/01-vectors-matrices/starter/starter.java
// Dependencies (Maven):
//   <dependency>
//     <groupId>org.apache.commons</groupId>
//     <artifactId>commons-math3</artifactId>
//     <version>3.6.1</version>
//   </dependency>
//
// Node: 1.1.1 — Vectors & Matrices: Core Operations
// Compile: javac -cp commons-math3-3.6.1.jar StarterAssignment.java
// Run:     java  -cp .:commons-math3-3.6.1.jar StarterAssignment

import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;

public class StarterAssignment {

    // -----------------------------------------------------------------------
    // Implemented helpers — provided for you, no changes needed
    // -----------------------------------------------------------------------

    /** Print a labelled section header. */
    static void printSeparator(String label) {
        System.out.println("==================================================");
        System.out.println(label);
        System.out.println("==================================================");
    }

    /** Print a pass/fail line comparing two double values within a tolerance. */
    static boolean verifyClose(String label, double actual, double expected, double atol) {
        boolean pass = Math.abs(actual - expected) <= atol;
        System.out.printf("  [%s] %s%n", pass ? "PASS" : "FAIL", label);
        return pass;
    }

    /** Print a pass/fail line comparing two matrices within a tolerance. */
    static boolean verifyMatrixClose(String label, double[][] actual,
                                     double[][] expected, double atol) {
        for (int i = 0; i < actual.length; i++) {
            for (int j = 0; j < actual[i].length; j++) {
                if (Math.abs(actual[i][j] - expected[i][j]) > atol) {
                    System.out.printf("  [FAIL] %s%n", label);
                    return false;
                }
            }
        }
        System.out.printf("  [PASS] %s%n", label);
        return true;
    }

    // -----------------------------------------------------------------------
    // Task 1 — Cosine Similarity
    // -----------------------------------------------------------------------

    /**
     * Return the cosine similarity between two double arrays of equal length.
     *
     * Steps:
     *   1. Compute the numerator as the sum of element-wise products (do NOT
     *      use ArrayRealVector.dotProduct — multiply manually in a loop).
     *   2. Compute the L2 norm of each vector (loop-based sqrt of sum of squares).
     *   3. Return numerator / (normA * normB).
     *
     * TODO: implement this method.
     */
    static double cosineSimilarity(double[] a, double[] b) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this");
    }

    // -----------------------------------------------------------------------
    // Task 2 — Row-Wise L2 Normalisation
    // -----------------------------------------------------------------------

    /**
     * Return a new matrix where every row of X has unit L2 norm.
     * Rows whose norm is zero are left unchanged (copied as-is).
     *
     * Steps:
     *   1. Allocate a new double[][] with the same dimensions as X.
     *   2. For each row, compute its L2 norm.
     *   3. If norm > 0, divide each element by the norm; otherwise copy unchanged.
     *
     * TODO: implement this method.
     */
    static double[][] rowNormalize(double[][] X) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this");
    }

    // -----------------------------------------------------------------------
    // Task 3 — Batch Feature Projection
    // -----------------------------------------------------------------------

    /**
     * Project each row of feature matrix X (n×d) onto weight vector (d,).
     * Return a double[] of length n where element i is the dot product of
     * row i of X with weights.
     *
     * Use RealMatrix.multiply(RealVector) or equivalent — no explicit loops.
     *
     * TODO: implement this method.
     */
    static double[] batchProject(double[][] X, double[] weights) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this");
    }

    // -----------------------------------------------------------------------
    // Task 4 — Power Iteration for Dominant Eigenpair
    // -----------------------------------------------------------------------

    /** Simple container for an eigenpair result. */
    record EigenPair(double eigenvalue, double[] eigenvector) {}

    /**
     * Approximate the dominant eigenpair of symmetric square matrix M.
     *
     * Algorithm:
     *   1. Initialise vector v = {1.0, 1.0, ..., 1.0} (length = M rows).
     *   2. Normalise v to unit L2 norm.
     *   3. For nIter steps:
     *        v = M × v   (matrix-vector multiply via RealMatrix.operate)
     *        normalise v to unit L2 norm
     *   4. Compute eigenvalue as the Rayleigh quotient: λ = vᵀ M v
     *        (i.e. dotProduct(v, M.operate(v))).
     *   5. Return new EigenPair(eigenvalue, v.toArray()).
     *
     * TODO: implement this method.
     */
    static EigenPair powerIteration(double[][] mData, int nIter) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this");
    }

    // -----------------------------------------------------------------------
    // Task 5 — Covariance Matrix
    // -----------------------------------------------------------------------

    /**
     * Compute the (d×d) sample covariance matrix from data matrix X (n×d).
     *
     * Steps:
     *   1. Compute the column mean vector (length d).
     *   2. Subtract the mean from each row to produce X_centred (n×d).
     *   3. Return (X_centredᵀ × X_centred) / (n - 1) as a double[][].
     *
     * Do NOT use Apache Commons CovarianceMatrix or similar utility.
     * Use only RealMatrix operations (multiply, transpose, scalarMultiply).
     *
     * TODO: implement this method.
     */
    static double[][] covarianceMatrix(double[][] X) {
        // TODO: implement this
        throw new UnsupportedOperationException("TODO: implement this");
    }

    // -----------------------------------------------------------------------
    // main — runs all tasks in order with labelled output
    // -----------------------------------------------------------------------

    public static void main(String[] args) {

        // ------------------------------------------------------------------
        // Task 1
        // ------------------------------------------------------------------
        printSeparator("Task 1 — Cosine Similarity");
        double[] a = {1.0, 2.0, 3.0};
        double[] b = {4.0, 5.0, 6.0};
        double sim = cosineSimilarity(a, b);
        System.out.printf("  a                      : [1.0, 2.0, 3.0]%n");
        System.out.printf("  b                      : [4.0, 5.0, 6.0]%n");
        System.out.printf("  cosineSimilarity(a, b) : %.4f%n", sim);
        verifyClose("Expected ≈ 0.9746", sim, 0.9746, 1e-4);
        System.out.println();

        // ------------------------------------------------------------------
        // Task 2
        // ------------------------------------------------------------------
        printSeparator("Task 2 — Row-Wise L2 Normalisation");
        double[][] Xnorm = {{3.0, 4.0}, {1.0, 0.0}, {0.0, 5.0}};
        double[][] normalised = rowNormalize(Xnorm);
        System.out.println("  Normalised rows:");
        for (int i = 0; i < normalised.length; i++) {
            double norm = 0;
            for (double v : normalised[i]) norm += v * v;
            norm = Math.sqrt(norm);
            System.out.printf("    row %d: [%.4f, %.4f]  norm=%.4f%n",
                    i, normalised[i][0], normalised[i][1], norm);
        }
        System.out.println();

        // ------------------------------------------------------------------
        // Task 3
        // ------------------------------------------------------------------
        printSeparator("Task 3 — Batch Feature Projection");
        double[][] Xproj  = {{0.25, 0.80}, {0.60, 0.45}, {0.90, 0.10}};
        double[]   weights = {0.7, 0.3};
        double[]   projections = batchProject(Xproj, weights);
        System.out.println("  Projections (expected: 0.4150, 0.5550, 0.6600):");
        for (int i = 0; i < projections.length; i++) {
            System.out.printf("    User %d: %.4f%n", i, projections[i]);
        }
        System.out.println();

        // ------------------------------------------------------------------
        // Task 4
        // ------------------------------------------------------------------
        printSeparator("Task 4 — Power Iteration (dominant eigenpair)");
        double[][] M = {{4.0, 2.0}, {2.0, 3.0}};
        EigenPair ep = powerIteration(M, 100);
        System.out.printf("  Dominant eigenvalue  : %.4f%n", ep.eigenvalue());
        System.out.printf("  Dominant eigenvector : [%.4f, %.4f]%n",
                ep.eigenvector()[0], ep.eigenvector()[1]);
        verifyClose("Eigenvalue ≈ 5.5616 (±0.001)", ep.eigenvalue(), 5.5616, 1e-3);
        // Verify M·v ≈ λ·v
        RealMatrix Mmat = new Array2DRowRealMatrix(M, false);
        RealVector v    = new ArrayRealVector(ep.eigenvector(), false);
        RealVector Mv   = Mmat.operate(v);
        RealVector lv   = v.mapMultiply(ep.eigenvalue());
        boolean match   = Mv.subtract(lv).getNorm() < 1e-6;
        System.out.printf("  Verification M·v ≈ λ·v : %b%n", match);
        System.out.println();

        // ------------------------------------------------------------------
        // Task 5
        // ------------------------------------------------------------------
        printSeparator("Task 5 — Covariance Matrix");
        double[][] Xcov = {
            {2.0, 0.0, 4.0},
            {3.0, 1.0, 5.0},
            {4.0, 2.0, 6.0},
            {5.0, 3.0, 7.0}
        };
        double[][] cov = covarianceMatrix(Xcov);
        System.out.println("  Covariance matrix:");
        for (double[] row : cov) {
            System.out.printf("    [%.4f, %.4f, %.4f]%n", row[0], row[1], row[2]);
        }
        // Expected: all cells ≈ 1.6667
        double[][] expectedCov = {
            {5.0/3.0, 5.0/3.0, 5.0/3.0},
            {5.0/3.0, 5.0/3.0, 5.0/3.0},
            {5.0/3.0, 5.0/3.0, 5.0/3.0}
        };
        verifyMatrixClose("Covariance matches expected (±1e-10)", cov, expectedCov, 1e-10);
        System.out.println();
    }
}

File: modules/01-math-foundations/01-linear-algebra/01-vectors-matrices/starter/starter.py
# Dependencies: numpy>=1.24
# Node: 1.1.1 — Vectors & Matrices: Core Operations
# Run: python starter.py

import numpy as np
from typing import Tuple


# ---------------------------------------------------------------------------
# Implemented helpers — provided for you, no changes needed
# ---------------------------------------------------------------------------

def print_separator(label: str) -> None:
    """Print a labelled section header."""
    print("=" * 50)
    print(f"{label}")
    print("=" * 50)


def load_sample_data() -> dict:
    """Return fixed arrays used across all tasks."""
    return {
        # Task 1
        "a": np.array([1.0, 2.0, 3.0]),
        "b": np.array([4.0, 5.0, 6.0]),
        # Task 2
        "X_norm": np.array([[3.0, 4.0],
                            [1.0, 0.0],
                            [0.0, 5.0]]),
        # Task 3
        "X_proj": np.array([[0.25, 0.80],
                            [0.60, 0.45],
                            [0.90, 0.10]]),
        "weights": np.array([0.7, 0.3]),
        # Task 4
        "M": np.array([[4.0, 2.0],
                       [2.0, 3.0]]),
        # Task 5
        "X_cov": np.array([[2.0, 0.0, 4.0],
                           [3.0, 1.0, 5.0],
                           [4.0, 2.0, 6.0],
                           [5.0, 3.0, 7.0]]),
    }


def verify_close(label: str, actual: np.ndarray, expected: np.ndarray,
                 atol: float = 1e-10) -> bool:
    """Print pass/fail for an array comparison and return the bool result."""
    match = np.allclose(actual, expected, atol=atol)
    status = "PASS" if match else "FAIL"
    print(f"  [{status}] {label}")
    return match


# ---------------------------------------------------------------------------
# Task 1 — Cosine Similarity
# ---------------------------------------------------------------------------

def cosine_similarity(a: np.ndarray, b: np.ndarray) -> float:
    """Return the cosine similarity between two 1-D vectors.

    Compute the numerator as element-wise multiply then sum (no np.dot).
    Divide by the product of the L2 norms.
    """
    raise NotImplementedError("TODO: implement this")


# ---------------------------------------------------------------------------
# Task 2 — Row-Wise L2 Normalisation
# ---------------------------------------------------------------------------

def row_normalize(X: np.ndarray) -> np.ndarray:
    """Return a copy of X where every row has unit L2 norm.

    Rows with zero norm are left unchanged.
    """
    raise NotImplementedError("TODO: implement this")


# ---------------------------------------------------------------------------
# Task 3 — Batch Feature Projection
# ---------------------------------------------------------------------------

def batch_project(X: np.ndarray, weights: np.ndarray) -> np.ndarray:
    """Project each row of X onto weights using the @ operator.

    Args:
        X:       shape (n, d) — feature matrix, one sample per row
        weights: shape (d,)   — projection weights

    Returns:
        shape (n,) — one scalar projection per sample
    """
    raise NotImplementedError("TODO: implement this")


# ---------------------------------------------------------------------------
# Task 4 — Power Iteration for Dominant Eigenpair
# ---------------------------------------------------------------------------

def power_iteration(M: np.ndarray, n_iter: int = 100) -> Tuple[float, np.ndarray]:
    """Approximate the dominant eigenpair of symmetric square matrix M.

    Algorithm:
        1. Seed np.random with 42, draw a random starting vector v of length M.shape[0].
        2. Normalise v to unit length.
        3. For n_iter steps: v = M @ v, then re-normalise v.
        4. Compute eigenvalue as the Rayleigh quotient: lambda = v.T @ M @ v.
        5. Return (lambda, v).

    Args:
        M:      square symmetric matrix, shape (n, n)
        n_iter: number of power iteration steps

    Returns:
        (dominant_eigenvalue: float, dominant_eigenvector: np.ndarray of shape (n,))
    """
    raise NotImplementedError("TODO: implement this")


# ---------------------------------------------------------------------------
# Task 5 — Covariance Matrix
# ---------------------------------------------------------------------------

def covariance_matrix(X: np.ndarray) -> np.ndarray:
    """Compute the (d, d) sample covariance matrix from data matrix X (n, d).

    Steps:
        1. Subtract the column mean from each row to centre the data.
        2. Return (X_centred.T @ X_centred) / (n - 1).

    Do NOT call np.cov.
    """
    raise NotImplementedError("TODO: implement this")


# ---------------------------------------------------------------------------
# main — runs all tasks in order with labelled output
# ---------------------------------------------------------------------------

def main() -> None:
    data = load_sample_data()

    # ------------------------------------------------------------------
    # Task 1
    # ------------------------------------------------------------------
    print_separator("Task 1 — Cosine Similarity")
    a, b = data["a"], data["b"]
    sim = cosine_similarity(a, b)
    print(f"  a                      : {a}")
    print(f"  b                      : {b}")
    print(f"  cosine_similarity(a,b) : {sim:.4f}")
    print()

    # ------------------------------------------------------------------
    # Task 2
    # ------------------------------------------------------------------
    print_separator("Task 2 — Row-Wise L2 Normalisation")
    X_norm = data["X_norm"]
    X_out = row_normalize(X_norm)
    print(f"  Input X:\n  {X_norm}")
    print(f"  Normalized rows:\n  {X_out}")
    row_norms = np.linalg.norm(X_out, axis=1)
    print(f"  Row norms after normalisation: {row_norms}")
    print()

    # ------------------------------------------------------------------
    # Task 3
    # ------------------------------------------------------------------
    print_separator("Task 3 — Batch Feature Projection")
    X_proj, weights = data["X_proj"], data["weights"]
    projections = batch_project(X_proj, weights)
    print(f"  Feature matrix X (3 users × 2 features):\n  {X_proj}")
    print(f"  Weights: {weights}")
    print(f"  Projections: {np.round(projections, 4)}")
    print()

    # ------------------------------------------------------------------
    # Task 4
    # ------------------------------------------------------------------
    print_separator("Task 4 — Power Iteration (dominant eigenpair)")
    M = data["M"]
    lam, v = power_iteration(M, n_iter=100)
    Mv = M @ v
    lv = lam * v
    print(f"  M:\n  {M}")
    print(f"  Dominant eigenvalue  : {lam:.4f}")
    print(f"  Dominant eigenvector : {np.round(v, 4)}")
    print(f"  Verification M·v ≈ λ·v : {np.allclose(Mv, lv, atol=1e-6)}")
    print()

    # ------------------------------------------------------------------
    # Task 5
    # ------------------------------------------------------------------
    print_separator("Task 5 — Covariance Matrix")
    X_cov = data["X_cov"]
    cov = covariance_matrix(X_cov)
    expected_cov = np.cov(X_cov, rowvar=False)
    print(f"  X (4 samples × 3 features):\n  {X_cov}")
    print(f"  Covariance matrix:\n  {np.round(cov, 4)}")
    verify_close("Matches np.cov", cov, expected_cov, atol=1e-10)
    print()


if __name__ == "__main__":
    main()
